### PR TITLE
Concurrent frontend runs

### PR DESCRIPTION
**Pull request by konradweiss**
Runs frontends for for file processing in parallel and merges the resulting CPG-ASTs of the runs before running the passes

### CODE REVIEW

**Review by oxisto — CHANGES_REQUESTED** [review]
Tests in CXXLanguageFrontend seem to hang after the merge

**Comment by oxisto** [comment]
The problem seems to be in `SubGraphWalker::activateTypes`. I am not quite what this function actually does. It is not documented and why this function is part of the subgraph walker.

**Comment by Masrepus** [comment]
As its name suggests, the type information that is initially only cached by the type manager is activated once the frontends are done. This function just walks the AST from leaf to root and sets the type where applicable

File: cpg-library/src/main/java/de/fraunhofer/aisec/cpg/TranslationConfiguration.java
Comment (Masrepus) [reply cpg-library/src/main/java/de/fraunhofer/aisec/cpg/TranslationConfiguration.java:447]: Yes, that's just a typo in the docs

File: cpg-library/src/main/java/de/fraunhofer/aisec/cpg/graph/declarations/FieldDeclaration.java
Comment (Masrepus) [reply cpg-library/src/main/java/de/fraunhofer/aisec/cpg/graph/declarations/FieldDeclaration.java:132]: I experimented with different ways of preventing this and the way this is currently done was the most stable one that didn't need that many changes

File: cpg-library/src/main/java/de/fraunhofer/aisec/cpg/meta/Edge.java
Comment (Masrepus) [reply cpg-library/src/main/java/de/fraunhofer/aisec/cpg/meta/Edge.java:32]: That's more of a general view on edges that makes it straightforward to compare them. PropertyEdge is way more complex and isn't really suited for graph comparison that I need for my purposes. You can of course remove that from this PR as this is not relevant for concurrent frontends

**Comment by oxisto** [comment]
> As its name suggests, the type information that is initially only cached by the type manager is activated once the frontends are done. This function just walks the AST from leaf to root and sets the type where applicable

Ok, understood. I would propose moving this to the type manager then. I guess that `setType` loops indefinitely somewhere after my merge changes :(

**Comment by Masrepus** [comment]
Sure, I don't mind where this function sits exactly.

Hmm I'll see if I find the time to debug it a bit in the coming days, maybe I'll find the issue

File: cpg-library/src/main/java/de/fraunhofer/aisec/cpg/meta/Edge.java
Comment (oxisto) [reply cpg-library/src/main/java/de/fraunhofer/aisec/cpg/meta/Edge.java:32]: > That's more of a general view on edges that makes it straightforward to compare them. PropertyEdge is way more complex and isn't really suited for graph comparison that I need for my purposes. You can of course remove that from this PR as this is not relevant for concurrent frontends

Aaah, understand. Yes, so I think to not interfere with your thesis it would probably be good if you branch of this existing branch, maybe even before my merge? Then let's focus on concurrency here

Comment (Masrepus) [reply cpg-library/src/main/java/de/fraunhofer/aisec/cpg/meta/Edge.java:32]: That's not an issue, I already branched off this one 👍

**Comment by oxisto** [comment]
@konradweiss before we merge this in, let's please discuss if this is possible without deactivating / changing the type system.

**Comment by oxisto** [comment]
> Kudos, SonarCloud Quality Gate passed!    Quality Gate passed
> 
> [Bug](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=VULNERABILITY)
> [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=SECURITY_HOTSPOT)
> [Code Smell](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=CODE_SMELL) [17 Code Smells](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=Fraunhofer-AISEC_cpg&pullRequest=444&metric=new_coverage&view=list) [93.5% Coverage](https://sonarcloud.io/component_measures?id=Fraunhofer-AISEC_cpg&pullRequest=444&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=Fraunhofer-AISEC_cpg&pullRequest=444&metric=new_duplicated_lines_density&view=list) [13.8% Duplication](https://sonarcloud.io/component_measures?id=Fraunhofer-AISEC_cpg&pullRequest=444&metric=new_duplicated_lines_density&view=list)

Can somebody have a look at the code smells, especially this one:

https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&open=AXqrDldWIcI4eL9dU-Et&pullRequest=444&resolved=false&types=CODE_SMELL

**Comment by konradweiss** [comment]
> > Kudos, SonarCloud Quality Gate passed!    Quality Gate passed
> > [Bug](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=BUG)
> > [Vulnerability](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=VULNERABILITY)
> > [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=SECURITY_HOTSPOT)
> > [Code Smell](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=CODE_SMELL) [17 Code Smells](https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&pullRequest=444&resolved=false&types=CODE_SMELL)
> > [](https://sonarcloud.io/component_measures?id=Fraunhofer-AISEC_cpg&pullRequest=444&metric=new_coverage&view=list) [93.5% Coverage](https://sonarcloud.io/component_measures?id=Fraunhofer-AISEC_cpg&pullRequest=444&metric=new_coverage&view=list)
> > [](https://sonarcloud.io/component_measures?id=Fraunhofer-AISEC_cpg&pullRequest=444&metric=new_duplicated_lines_density&view=list) [13.8% Duplication](https://sonarcloud.io/component_measures?id=Fraunhofer-AISEC_cpg&pullRequest=444&metric=new_duplicated_lines_density&view=list)
> 
> Can somebody have a look at the code smells, especially this one:
> 
> https://sonarcloud.io/project/issues?id=Fraunhofer-AISEC_cpg&open=AXqrDldWIcI4eL9dU-Et&pullRequest=444&resolved=false&types=CODE_SMELL

I guess it does not like usage of the binary operator | instead of the boolean operator |. The results in the mapped cases is the same, as the | is only used for values that do not have to be evaluated and the distinction to the || Operator does not hold. I will jsut replace it with || and look at some of the other smells

**Review by oxisto — APPROVED** [review]
LGTM for now. In the future, we might need to finally tackle optimising the type manager / type system.

**Pull request merged**
